### PR TITLE
Add validation for number of arguments provided.

### PR DESCRIPTION
**Pull request by Makiv1**
parse_full_telecommand() now validates that the correct number of arguments are provided. Returns 100 if the wrong number of arguments are provided. Corresponding unit tests have been added.

### CODE REVIEW

File: firmware/Core/Src/telecommands/telecommand_parser.c [hunk 1]
```diff
@@ -279,7 +279,38 @@ uint8_t TCMD_parse_full_telecommand(const char tcmd_str[], TCMD_TelecommandChann
         return 90;
     }
 
-    // Reached the end of the telecommand parsing. Thus, success. Fill the output struct.
+    // ensure the corrent number of args are provided
+    int32_t num_args_expected = TCMD_telecommand_definitions[tcmd_idx].number_of_args;
+
+    char error_message[80];
+    snprintf(error_message, 80, "Error: TCMD_parse_full_telecommand: this telecommand acepts %d arguments.\n", num_args_expected);
```
Comment (NuclearTea) [inline firmware/Core/Src/telecommands/telecommand_parser.c:286]: 2 things:

1. I think you should be able to say the actual name of the telecommand here.
2. Spelling mistake for accept.

File: firmware/Core/Src/unit_tests/test_telecommand_parser.c
Comment (NuclearTea) [inline firmware/Core/Src/unit_tests/test_telecommand_parser.c:1]: When testing a failing case, comment what you are testing, for example: 

// Testing telecommand which takes 0 args, providing 1 arg

Also, you are checking whether the function returns != 0 but you explicitly declare the function to return 100 if something goes wrong, so you should test that this returns 100. Otherwise, something else could go wrong, and it returns something else (like 10), you don't know whether that was because of your inputs, or something else.

Otherwise, nice selection of tests. Seem to be pretty complete

**Review by NuclearTea — COMMENTED** [review]
Nice work overall! 

Remember to rebase with main.

File: firmware/Core/Src/telecommands/telecommand_parser.c (hunk 1 above)
Comment (DeflateAwning) [reply firmware/Core/Src/telecommands/telecommand_parser.c:286]: 3. Use `sizeof(error_message)` as the second arg to `error_message` (best practice to avoid repeating magic numbers).

File: firmware/Core/Inc/unit_tests/test_telecommand_parser.h [hunk 2]
```diff
@@ -12,4 +12,6 @@ uint8_t TEST_EXEC__TCMD_get_suffix_tag_uint64();
 
 uint8_t TEST_EXEC__TCMD_ascii_to_double();
 
+uint8_t TEST_EXEC_TCMD_TCMD_parse_full_telecommand();
```
Comment (DeflateAwning) [inline firmware/Core/Inc/unit_tests/test_telecommand_parser.h:15]: Thanks for working on this :) 

Function name should be `TEST_EXEC_TCMD_parse_full_telecommand`

File: firmware/Core/Src/telecommands/telecommand_parser.c [hunk 3]
```diff
@@ -279,7 +279,38 @@ uint8_t TCMD_parse_full_telecommand(const char tcmd_str[], TCMD_TelecommandChann
...
+    if ( num_args_expected == 0 && arg_len > 0) {
+        DEBUG_uart_print_str(error_message);
+        return 100;
+    }
+    if ( num_args_expected > 0 && arg_len == 0) {
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/telecommand_parser.c:291]: This code is duplicated.

File: firmware/Core/Src/telecommands/telecommand_parser.c [hunk 4]
```diff
@@ -279,7 +279,38 @@ uint8_t TCMD_parse_full_telecommand(const char tcmd_str[], TCMD_TelecommandChann
...
+        DEBUG_uart_print_str(error_message);
+        return 100;
+    }
+
+    char remaining_args_to_parse[arg_len + 1];
+    memcpy(remaining_args_to_parse, &args_str_no_parens, arg_len +1);
+
+    for (int32_t i = 0; i < num_args_expected - 1; i++) {
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/telecommand_parser.c:299]: This logic feels convoluted. Why not just count the number of commas, and compare it to `num_args_expected-1`?

File: firmware/Core/Src/unit_tests/test_telecommand_parser.c [hunk 5]
```diff
@@ -122,3 +122,16 @@ uint8_t TEST_EXEC__TCMD_ascii_to_double() {
 
     return 0;
 }
+
+uint8_t TEST_EXEC_TCMD_TCMD_parse_full_telecommand() {
+    TCMD_parsed_tcmd_to_execute_t parsed_tcmd;
+    TEST_ASSERT_TRUE(TCMD_parse_full_telecommand("CTS1+hello_world()!", TCMD_TelecommandChannel_DEBUG_UART, &parsed_tcmd) == 0);
```
Comment (DeflateAwning) [inline firmware/Core/Src/unit_tests/test_telecommand_parser.c:128]: In this test case, please validate that the members of the struct are set correctly as well, as best as you can.

**Review by DeflateAwning — CHANGES_REQUESTED** [review]
Very awesome, thanks for working on this!

**Comment by DeflateAwning** [comment]
Let me know when this is ready for another review!

**Comment by Makiv1** [comment]
Should be ready for another review :)

File: firmware/Core/Src/telecommands/telecommand_parser.c [hunk 6]
```diff
@@ -279,7 +279,31 @@ uint8_t TCMD_parse_full_telecommand(const char tcmd_str[], TCMD_TelecommandChann
...
 
-    // Reached the end of the telecommand parsing. Thus, success. Fill the output struct.
+    // ensure the corrent number of args are provided
+    int32_t num_args_expected = TCMD_telecommand_definitions[tcmd_idx].number_of_args;
+    char* tcmd_name = TCMD_telecommand_definitions[tcmd_idx].tcmd_name;
+
+    char error_message[130];
+    snprintf(error_message, sizeof(error_message), "Error: TCMD_parse_full_telecommand: %s() accepts %d argument(s).\n", tcmd_name, num_args_expected);
+    // if ( num_args_expected == 0 && arg_len > 0) {
+    //     DEBUG_uart_print_str(error_message);
```
Comment (NuclearTea) [inline firmware/Core/Src/telecommands/telecommand_parser.c:289]: Just a quick thing I notice, need to remove this commented code

File: firmware/.gitignore [hunk 7]
```diff
@@ -19,3 +19,5 @@
 **/STM32Make.make
 **/.vscode/settings.json
 **/.vscode/c_cpp_properties.json
+
+./vscode/launch.json
```
Comment (DeflateAwning) [inline firmware/.gitignore:23]: Please revert this change, unless you can explain it :)

File: firmware/Core/Src/telecommands/telecommand_parser.c [hunk 8]
```diff
@@ -279,7 +279,31 @@ uint8_t TCMD_parse_full_telecommand(const char tcmd_str[], TCMD_TelecommandChann
...
+    //     return 100;
+    // }
+    int32_t num_commas = 0;
+    for (int32_t i = 0; i < arg_len + 1; i++) {
+        if (args_str_no_parens[i] == ',') {
+            num_commas++;
+        }    
+    }
+    int32_t takes_no_arguments_and_none_provided = num_args_expected == 0 && arg_len == 0;
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/telecommand_parser.c:298]: Pretty sure that the validation logic should just be:
```
    (num_args_expected == 0 && arg_len == 0) || (num_commas == (num_args_expected-1))
```

Thoughts? Am I missing something?

File: firmware/Core/Src/unit_tests/test_telecommand_parser.c [hunk 9]
```diff
@@ -122,3 +122,25 @@ uint8_t TEST_EXEC__TCMD_ascii_to_double() {
...
+    TEST_ASSERT_TRUE(strcmp(parsed_tcmd.args_str_no_parens, "") == 0);
+    TEST_ASSERT_TRUE(parsed_tcmd.tcmd_channel == TCMD_TelecommandChannel_DEBUG_UART);
+    TEST_ASSERT_TRUE(strcmp(TCMD_telecommand_definitions[parsed_tcmd.tcmd_idx].tcmd_name, "hello_world") == 0);
+
+    // Tesing incorrect usage of 0 argument telecommand
+    TEST_ASSERT_TRUE(TCMD_parse_full_telecommand("CTS1+hello_world(a)!", TCMD_TelecommandChannel_DEBUG_UART, &parsed_tcmd) == 100);
+    TEST_ASSERT_TRUE(TCMD_parse_full_telecommand("CTS1+hello_world(a,1)!", TCMD_TelecommandChannel_DEBUG_UART, &parsed_tcmd) == 100);
+
+    // Testing correct usage of 1 argument telecommand 
+    TEST_ASSERT_TRUE(TCMD_parse_full_telecommand("CTS1+set_system_time(10)!", TCMD_TelecommandChannel_DEBUG_UART, &parsed_tcmd) == 0);
```
Comment (DeflateAwning) [inline firmware/Core/Src/unit_tests/test_telecommand_parser.c:139]: For this "success" unit test, please check struct fields.

File: firmware/Core/Src/telecommands/telecommand_parser.c [hunk 10]
```diff
@@ -279,7 +279,31 @@ uint8_t TCMD_parse_full_telecommand(const char tcmd_str[], TCMD_TelecommandChann
         return 90;
     }
 
-    // Reached the end of the telecommand parsing. Thus, success. Fill the output struct.
+    // ensure the corrent number of args are provided
+    int32_t num_args_expected = TCMD_telecommand_definitions[tcmd_idx].number_of_args;
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/telecommand_parser.c:283]: `int16_t` would be more appropriate (or even `int8_t`). Please use `const` keyword to mark variable as immutable.

File: firmware/Core/Src/unit_tests/test_telecommand_parser.c [hunk 11]
```diff
@@ -122,3 +122,25 @@ uint8_t TEST_EXEC__TCMD_ascii_to_double() {
...
+    TEST_ASSERT_TRUE(TCMD_parse_full_telecommand("CTS1+hello_world(a)!", TCMD_TelecommandChannel_DEBUG_UART, &parsed_tcmd) == 100);
+    TEST_ASSERT_TRUE(TCMD_parse_full_telecommand("CTS1+hello_world(a,1)!", TCMD_TelecommandChannel_DEBUG_UART, &parsed_tcmd) == 100);
+
+    // Testing correct usage of 1 argument telecommand 
+    TEST_ASSERT_TRUE(TCMD_parse_full_telecommand("CTS1+set_system_time(10)!", TCMD_TelecommandChannel_DEBUG_UART, &parsed_tcmd) == 0); 
+    
+    // Tesing incorrect usage of 1 argument telecommand
+    TEST_ASSERT_TRUE(TCMD_parse_full_telecommand("CTS1+set_system_time()!", TCMD_TelecommandChannel_DEBUG_UART, &parsed_tcmd) == 100);
+    TEST_ASSERT_TRUE(TCMD_parse_full_telecommand("CTS1+set_system_time(10,1)!", TCMD_TelecommandChannel_DEBUG_UART, &parsed_tcmd) == 100);
+
```
Comment (DeflateAwning) [inline firmware/Core/Src/unit_tests/test_telecommand_parser.c:144]: Please add the following test cases for which the validation should fail. I _think_ they're all caught right now, but still a good check.

```
CTS1+hello_world(,)!

// Edge cases: commas at starts and ends
CTS1+set_system_time(123,)!
CTS1+set_system_time(,123)!

// flash_read_hex is a 3-arg function
CTS1+flash_read_hex(0,0,0)! // Expected success
CTS1+flash_read_hex(0,0,0,)! // Expected fail
CTS1+flash_read_hex(,0,0,0)! // Expected fail
CTS1+flash_read_hex(0,0,0,)! // Expected fail
CTS1+flash_read_hex(0,0)! // Expected fail
```

File: firmware/Core/Src/telecommands/telecommand_parser.c [hunk 12]
```diff
@@ -279,7 +279,31 @@ uint8_t TCMD_parse_full_telecommand(const char tcmd_str[], TCMD_TelecommandChann
...
+    char* tcmd_name = TCMD_telecommand_definitions[tcmd_idx].tcmd_name;
+
+    char error_message[130];
+    snprintf(error_message, sizeof(error_message), "Error: TCMD_parse_full_telecommand: %s() accepts %d argument(s).\n", tcmd_name, num_args_expected);
+    // if ( num_args_expected == 0 && arg_len > 0) {
+    //     DEBUG_uart_print_str(error_message);
+    //     return 100;
+    // }
+    int32_t num_commas = 0;
+    for (int32_t i = 0; i < arg_len + 1; i++) {
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/telecommand_parser.c:293]: Please change upper loop bound to `i <= arg_len` (instead of `arg_len + 1`). It's more idomatic, and it decreases a potential place for an int wrap-around (not really likely here, but good to care about in general).

**Review by DeflateAwning — CHANGES_REQUESTED** [review]
Good improvements!

* Remove commented-out code.
* Ensure a clean build passes. Clean build currently has a warning.
* Rebase to main.
* See comments.

**Comment by DeflateAwning** [comment]
If you're getting tired of this task, feel free to let me know and I'd be happy to make the remaining changes, and/or you can get started on another issue!

File: firmware/Core/Src/telecommands/telecommand_parser.c (hunk 8 above)
Comment (Makiv1) [reply firmware/Core/Src/telecommands/telecommand_parser.c:298]: If a telecommand accepts 1 argument but no argument is given this validation logic would fail. Changing it to: `(num_args_expected == 0 && arg_len == 0) || (num_commas == (num_args_expected-1) && arg_len != 0)` should work though.

Comment (DeflateAwning) [reply firmware/Core/Src/telecommands/telecommand_parser.c]: Ah, great catch! I like that though; the three-part logic seemed to have overlapping cases, I think.

File: firmware/Core/Src/telecommands/telecommand_parser.c [hunk 13]
```diff
@@ -279,7 +279,26 @@ uint8_t TCMD_parse_full_telecommand(const char tcmd_str[], TCMD_TelecommandChann
...
+    snprintf(error_message, sizeof(error_message), "Error: TCMD_parse_full_telecommand: %s() accepts %d argument(s).\n", tcmd_name, num_args_expected);
+
+    int16_t num_commas = 0;
+    for (int32_t i = 0; i <= arg_len; i++) {
+        if (args_str_no_parens[i] == ',') {
+            num_commas++;
+        }    
+    }
+    
+    int8_t correct_number_of_args_provided = (num_args_expected == 0 && arg_len == 0) || (num_commas == (num_args_expected-1) && arg_len != 0);
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/telecommand_parser.c:296]: Ideally, this should be marked `const` to label it as immutable

**Review by DeflateAwning — DISMISSED** [review]
Solid! One optional fix (in a comment).

**Comment by DeflateAwning** [comment]
Also, if you make that change, feel free to add yourself as a contributor in the README!

**Review by DeflateAwning — APPROVED** [review]

**Review by DeflateAwning — APPROVED** [review]
Looks good! Please Squash and Merge this PR.

In generally, we prefer to rebase instead of merging `main` into your feature branch.
